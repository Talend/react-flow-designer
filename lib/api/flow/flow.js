'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.updateNode = exports.deleteNode = exports.addNode = exports.isNodeExist = undefined;

var _flow = require('lodash/flow');

var _flow2 = _interopRequireDefault(_flow);

var _curry = require('lodash/curry');

var _curry2 = _interopRequireDefault(_curry);

var _immutable = require('immutable');

var _immutable2 = _interopRequireDefault(_immutable);

var _throwInDev = require('../throwInDev');

var _ = require('./..');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * check if node exist in flow
 * @param {FlowState} state
 * @param {string} nodeId
 * @return {bool} true if node exist
 */
var isNodeExist = exports.isNodeExist = (0, _curry2.default)(function (state, nodeId) {
	return state.hasIn(['nodes', nodeId]);
});

var setOut = (0, _curry2.default)(function (nodeId, state) {
	return state.setIn(['out', nodeId], new _immutable2.default.Map());
});
var deleteOut = (0, _curry2.default)(function (nodeId, state) {
	return state.deleteIn(['out', nodeId]);
});

var setIn = (0, _curry2.default)(function (nodeId, state) {
	return state.setIn(['in', nodeId], new _immutable2.default.Map());
});
var deleteIn = (0, _curry2.default)(function (nodeId, state) {
	return state.deleteIn(['in', nodeId]);
});

var setChildrens = (0, _curry2.default)(function (nodeId, state) {
	return state.setIn(['childrens', nodeId], new _immutable2.default.Map());
});
var deleteChildrens = (0, _curry2.default)(function (nodeId, state) {
	return state.deleteIn(['childrens', nodeId]);
});

var setParents = (0, _curry2.default)(function (nodeId, state) {
	return state.setIn(['parents', nodeId], new _immutable2.default.Map());
});
var deleteParents = (0, _curry2.default)(function (nodeId, state) {
	return state.deleteIn(['parents', nodeId]);
});

/**
 * add a node to the flow
 * @param {FlowState} state
 * @param {NodeRecord} node
 * @return {FlowState}
 */
var addNode = exports.addNode = (0, _curry2.default)(function (state, node) {
	if (_.Node.isNodeElseThrow(node) && !isNodeExist(state, _.Node.getId(node))) {
		var nodeId = _.Node.getId(node);
		return (0, _flow2.default)([setOut(nodeId), setIn(nodeId), setChildrens(nodeId), setParents(nodeId)])(state.setIn(['nodes', _.Node.getId(node)], node));
	}
	(0, _throwInDev.throwInDev)('Node with id = ' + _.Node.getId(node) + ', already exist, can\'t create node.');
	return state;
});

/**
 * if exist remove a node from the flow
 * @param {FlowState} state
 * @param {string} nodeId
 * @return {FlowState}
 */
var deleteNode = exports.deleteNode = (0, _curry2.default)(function (state, nodeId) {
	if (isNodeExist(nodeId)) {
		return (0, _flow2.default)([deleteOut(nodeId), deleteIn(nodeId), deleteChildrens(nodeId), deleteParents(nodeId)])(state.deleteIn(['nodes', nodeId]));
	}
	return state;
});

/**
 * update a node
 * @param {FlowState} state
 * @param {string} nodeId
 * @param {NodeRecord} node
 * @return {FlowState}
 */
var updateNode = exports.updateNode = (0, _curry2.default)(function (state, nodeId, node) {
	if (_.Node.isNodeElseThrow(node) && (isNodeExist(state, _.Node.getId(node)), isNodeExist(nodeId))) {
		if (nodeId === _.Node.getId(node)) {
			return addNode(deleteNode(state, _.Node.getId(node)), node);
		}
		return state.setIn(['nodes', nodeId], node);
	}
	return state;
});