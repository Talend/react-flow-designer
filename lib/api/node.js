'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.Node = undefined;
exports.isNodeElseThrow = isNodeElseThrow;

var _curry = require('lodash/curry');

var _curry2 = _interopRequireDefault(_curry);

var _flow = require('lodash/flow');

var _flow2 = _interopRequireDefault(_flow);

var _indexOf = require('lodash/indexOf');

var _indexOf2 = _interopRequireDefault(_indexOf);

var _isString = require('lodash/isString');

var _isString2 = _interopRequireDefault(_isString);

var _upperFirst = require('lodash/upperFirst');

var _upperFirst2 = _interopRequireDefault(_upperFirst);

var _throwInDev = require('./throwInDev');

var _flowdesigner = require('../constants/flowdesigner.model');

var _position = require('./position');

var _size = require('./size');

var _data = require('./data');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This module is public and deal with Graph's object Nodes
 */
var positionSelector = ['graphicalAttributes', 'position'];
var sizeSelector = ['graphicalAttributes', 'nodeSize'];
var componentTypeSelector = ['graphicalAttributes', 'nodeType'];

/** in future properties should be removed from the react-flow-designer lib */
var FORBIDEN_GRAPHICAL_ATTRIBUTES = ['properties', 'position', 'nodeSize', 'nodeType'];

/**
 * Test if the first parameter is a NodeRecord instance
 * @param {NodeRecord} node
 * @returns {bool}
 * @throws
 */
function isNode(node) {
	if (node && node instanceof _flowdesigner.NodeRecord) {
		return true;
	}
	return false;
}

/**
 * Test if the first parameter is a NodeRecord, throw if not
 * @param {*} node
 * @returns {bool}
 * @throws
 */
function isNodeElseThrow(node) {
	var test = isNode(node);
	if (!test) {
		(0, _throwInDev.throwTypeError)('NodeRecord', node, 'node', 'Node');
	}
	return test;
}

/**
 * @param {NodeRecord} node
 * @returns {string}
 */
function getId(node) {
	if (isNodeElseThrow(node)) {
		return node.get('id');
	}
	return null;
}

/**
 * @param {string} id
 * @param {NodeRecord}
 * @returns {NodeRecord}
 */
var setId = (0, _curry2.default)(function (id, node) {
	if ((0, _isString2.default)(id) && isNodeElseThrow(node)) {
		return node.set('id', id);
	}
	(0, _throwInDev.throwInDev)('nodeId should be a string was given ' + (id && id.toString()));
	return node;
});

/**
 * @param {NodeRecord} node
 * @returns {PositionRecord}
 */
function getPosition(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(positionSelector);
	}
	return null;
}

/**
 * @param {PositionRecord} position
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
var setPosition = (0, _curry2.default)(function (position, node) {
	if ((0, _position.isPositionElseThrow)(position) && isNodeElseThrow(node)) {
		return node.setIn(positionSelector, position);
	}
	return node;
});

/**
 * @param {NodeRecord} node
 * @returns {PositionRecord}
 */
function getSize(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(sizeSelector);
	}
	return null;
}

/**
 * @param {SizeRecord} size
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
var setSize = (0, _curry2.default)(function (size, node) {
	if ((0, _size.isSizeElseThrow)(size) && isNodeElseThrow(node)) {
		return node.setIn(sizeSelector, size);
	}
	return node;
});

/**
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
function getComponentType(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(componentTypeSelector);
	}
	return null;
}

/**
 * @param {string} nodeType
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
var setComponentType = (0, _curry2.default)(function (nodeType, node) {
	if ((0, _isString2.default)(nodeType) && isNodeElseThrow(node)) {
		return node.setIn(componentTypeSelector, nodeType);
	}
	(0, _throwInDev.throwInDev)('nodeType should be a string was given ' + (nodeType && nodeType.toString()));
	return node;
});

/**
 * @param {String} key
 * @param {any} value
 * @param {nodeRecord} node
 * @returns {nodeRecord}
 */
var setData = (0, _curry2.default)(function (key, value, node) {
	if (isNodeElseThrow(node)) {
		return node.set('data', _data.Data.set(key, value, node.get('data')));
	}
	return node;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
var getData = (0, _curry2.default)(function (key, node) {
	if (isNodeElseThrow(node)) {
		return _data.Data.get(key, node.get('data'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
var hasData = (0, _curry2.default)(function (key, node) {
	if (isNodeElseThrow(node)) {
		return _data.Data.has(key, node.get('data'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
var deleteData = (0, _curry2.default)(function (key, node) {
	if (isNodeElseThrow(node)) {
		return node.set('data', _data.Data.delete(key, node.get('data')));
	}
	return node;
});

/**
 * given a key check if that key is white listed
 * @param {String} key
 * @returns {Bool}
 */
function isWhiteListAttribute(key) {
	if ((0, _indexOf2.default)(FORBIDEN_GRAPHICAL_ATTRIBUTES, key) === -1) {
		return true;
	}
	(0, _throwInDev.throwInDev)(key + ' is a protected value of the Node, please use get' + (0, _upperFirst2.default)(key) + ' set' + (0, _upperFirst2.default)(key) + ' from this module to make change on those values');
	return false;
}

/**
 * @param {String} key
 * @param {any} value
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
var setGraphicalAttribute = (0, _curry2.default)(function (key, value, node) {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return node.set('graphicalAttributes', _data.Data.set(key, value, node.get('graphicalAttributes')));
	}
	return node;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
var getGraphicalAttribute = (0, _curry2.default)(function (key, node) {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return _data.Data.get(key, node.get('graphicalAttributes'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
var hasGraphicalAttribute = (0, _curry2.default)(function (key, node) {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return _data.Data.has(key, node.get('graphicalAttributes'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
var deleteGraphicalAttribute = (0, _curry2.default)(function (key, node) {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return node.set('graphicalAttributes', _data.Data.delete(key, node.get('graphicalAttributes')));
	}
	return node;
});

/**
 * Create a new Node
 * @param {String} id
 * @param {PositionRecord} position
 * @param {SizeRecord} size
 * @param {String} componentType
 * @returns {NodeRecord}
 */
var create = (0, _curry2.default)(function (id, position, size, componentType) {
	return (0, _flow2.default)([setId(id), setPosition(position), setSize(size), setComponentType(componentType)])(new _flowdesigner.NodeRecord());
});

var Node = exports.Node = {
	create: create,
	isNode: isNode,
	getId: getId,
	getPosition: getPosition,
	setPosition: setPosition,
	getSize: getSize,
	setSize: setSize,
	getComponentType: getComponentType,
	setComponentType: setComponentType,
	setData: setData,
	getData: getData,
	hasData: hasData,
	deleteData: deleteData,
	setGraphicalAttribute: setGraphicalAttribute,
	getGraphicalAttribute: getGraphicalAttribute,
	hasGraphicalAttribute: hasGraphicalAttribute,
	deleteGraphicalAttribute: deleteGraphicalAttribute
};