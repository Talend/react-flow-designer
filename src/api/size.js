import { SizeRecord } from '../constants/flowdesigner.model';

export function isSizeRecord(size, doThrow = true) {
	if (size && size instanceof SizeRecord) {
		return true;
	}
	if (doThrow) {
		throw new Error(`Should be a sizeRecord was given ${size.toString()}`);
	}
	return false;
}
