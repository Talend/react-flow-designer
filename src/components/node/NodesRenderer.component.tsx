import PropTypes from 'prop-types';
import React from 'react';
import invariant from 'invariant';
import { mapOf } from 'react-immutable-proptypes';
import { Map } from 'immutable';

import { NodeType } from '../../constants/flowdesigner.proptypes';
import { NodeRecordMap, NodeRecord, Id, Position } from '../../customTypings/index.d';

type Props = {
	nodes: NodeRecordMap;
	nodeTypeMap: Map<string, any>;
	startMoveNodeTo: (nodeId: Id, nodePosition: string) => void;
	moveNodeTo: (nodeId: Id, nodePosition: Position) => void;
	moveNodeToEnd: (nodeId: Id, nodePosition: Position) => void;
	snapToGrid: boolean;
};

class NodesRenderer extends React.Component<Props> {
	static propTypes = {
		nodes: mapOf(NodeType).isRequired,
		nodeTypeMap: PropTypes.object.isRequired,
		startMoveNodeTo: PropTypes.func.isRequired,
		moveNodeTo: PropTypes.func.isRequired,
		moveNodeToEnd: PropTypes.func.isRequired,
		snapToGrid: PropTypes.bool.isRequired,
	};

	constructor(props: Props) {
		super(props);
		this.renderNode = this.renderNode.bind(this);
	}

	renderNode(node: NodeRecord) {
		const type = node.getNodeType();
		const ConcreteComponent = this.props.nodeTypeMap.get(type).component;
		if (!ConcreteComponent) {
			invariant(
				false,
				`<NodesRenderer />  the defined node type in your graph model hasn't been mapped into
				the dataflow configuration, check NodeType documentation`,
			);
		}
		return (
			<ConcreteComponent
				node={node}
				startMoveNodeTo={this.props.startMoveNodeTo}
				moveNodeTo={this.props.moveNodeTo}
				moveNodeToEnd={this.props.moveNodeToEnd}
				key={node.id}
				snapToGrid={this.props.snapToGrid}
			/>
		);
	}

	render() {
		return <g>{this.props.nodes.valueSeq().map(this.renderNode)}</g>;
	}
}

export default NodesRenderer;
