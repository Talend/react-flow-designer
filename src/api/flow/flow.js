import flow from 'lodash/flow';
import curry from 'lodash/curry';
import Immutable from 'immutable';

import { throwInDev } from '../throwInDev';
import { Node } from './..';

/**
 * check if node exist in flow
 * @param {FlowState} state
 * @param {string} nodeId
 * @return {bool} true if node exist
 */
export const isNodeExist = curry((state, nodeId) => state.hasIn(['nodes', nodeId]));

const setOut = curry((nodeId, state) => state.setIn(['out', nodeId], new Immutable.Map()));
const deleteOut = curry((nodeId, state) => state.deleteIn(['out', nodeId]));

const setIn = curry((nodeId, state) => state.setIn(['in', nodeId], new Immutable.Map()));
const deleteIn = curry((nodeId, state) => state.deleteIn(['in', nodeId]));

const setChildrens = curry((nodeId, state) =>
	state.setIn(['childrens', nodeId], new Immutable.Map()),
);
const deleteChildrens = curry((nodeId, state) => state.deleteIn(['childrens', nodeId]));

const setParents = curry((nodeId, state) => state.setIn(['parents', nodeId], new Immutable.Map()));
const deleteParents = curry((nodeId, state) => state.deleteIn(['parents', nodeId]));

/**
 * add a node to the flow
 * @param {FlowState} state
 * @param {NodeRecord} node
 * @return {FlowState}
 */
export const addNode = curry((state, node) => {
	if (Node.isNodeElseThrow(node) && !isNodeExist(state, Node.getId(node))) {
		const nodeId = Node.getId(node);
		return flow([setOut(nodeId), setIn(nodeId), setChildrens(nodeId), setParents(nodeId)])(
			state.setIn(['nodes', Node.getId(node)], node),
		);
	}
	throwInDev(`Node with id = ${Node.getId(node)}, already exist, can't create node.`);
	return state;
});

/**
 * if exist remove a node from the flow
 * @param {FlowState} state
 * @param {string} nodeId
 * @return {FlowState}
 */
export const deleteNode = curry((state, nodeId) => {
	if (isNodeExist(nodeId)) {
		return flow([
			deleteOut(nodeId),
			deleteIn(nodeId),
			deleteChildrens(nodeId),
			deleteParents(nodeId),
		])(state.deleteIn(['nodes', nodeId]));
	}
	return state;
});

/**
 * update a node
 * @param {FlowState} state
 * @param {string} nodeId
 * @param {NodeRecord} node
 * @return {FlowState}
 */
export const updateNode = curry((state, nodeId, node) => {
	if (Node.isNodeElseThrow(node) && (isNodeExist(state, Node.getId(node)), isNodeExist(nodeId))) {
		if (nodeId === Node.getId(node)) {
			return addNode(deleteNode(state, Node.getId(node)), node);
		}
		return state.setIn(['nodes', nodeId], node);
	}
	return state;
});
