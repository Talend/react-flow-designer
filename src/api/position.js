import curry from 'lodash/curry';
import flow from 'lodash/flow';
import isNumber from 'lodash/isNumber';

import throwInDev from './throwInDev';
import { PositionRecord } from '../constants/flowdesigner.model';

function isPosition(position) {
	if (position && position instanceof PositionRecord) {
		return true;
	}
	return false;
}

export function isPositionElseThrow(position) {
	const test = isPosition(position);
	if (!test) {
		throwInDev(
			`position should be a positionRecord was given ${position &&
				position.toString()}, you should use Position module functions to create and transform Positions`,
		);
	}
	return test;
}

function getXCoordinate(position) {
	if (isPositionElseThrow(position)) {
		return position.get('x');
	}
	return null;
}

const setXCoordinate = curry((x, position) => {
	if (isPositionElseThrow(position) && isNumber(x)) {
		return position.set('x', x);
	}
	throwInDev(`x should be a number was given ${x && x.toString()} of type ${typeof x}`);
	return position;
});

function getYCoordinate(position) {
	if (isPositionElseThrow(position)) {
		return position.get('y');
	}
	return null;
}

const setYCoordinate = curry((y, position) => {
	if (isPositionElseThrow(position) && isNumber(y)) {
		return position.set('y', y);
	}
	throwInDev(`y should be a number was given ${y && y.toString()} of type ${typeof y}`);
	return position;
});

const create = curry((x, y) => flow([setXCoordinate(x), setYCoordinate(y)])(new PositionRecord()));

export const Position = {
	create,
	isPosition,
	getXCoordinate,
	setXCoordinate,
	getYCoordinate,
	setYCoordinate,
};
