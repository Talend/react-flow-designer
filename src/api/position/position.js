import curry from 'lodash/curry';
import flow from 'lodash/flow';
import isNumber from 'lodash/isNumber';

import { throwInDev, throwTypeError } from '../throwInDev';
import { PositionRecord } from '../../constants/flowdesigner.model';

/**
 * @typedef {Immutable.Record} PositionRecord
 */

export function isPosition(position) {
	if (position && position instanceof PositionRecord) {
		return true;
	}
	return false;
}

export function isPositionElseThrow(position) {
	const test = isPosition(position);
	if (!test) {
		throwTypeError('PositionRecord', position, 'position', 'Position');
	}
	return test;
}

export function getXCoordinate(position) {
	if (isPositionElseThrow(position)) {
		return position.get('x');
	}
	return null;
}

export const setXCoordinate = curry((x, position) => {
	if (isPositionElseThrow(position) && isNumber(x)) {
		return position.set('x', x);
	}
	throwInDev(`x should be a number was given ${x && x.toString()} of type ${typeof x}`);
	return position;
});

export function getYCoordinate(position) {
	if (isPositionElseThrow(position)) {
		return position.get('y');
	}
	return null;
}

export const setYCoordinate = curry((y, position) => {
	if (isPositionElseThrow(position) && isNumber(y)) {
		return position.set('y', y);
	}
	throwInDev(`y should be a number was given ${y && y.toString()} of type ${typeof y}`);
	return position;
});

export const create = curry((x, y) => flow([setXCoordinate(x), setYCoordinate(y)])(new PositionRecord()));
