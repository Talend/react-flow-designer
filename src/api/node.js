/**
 * This module is public and deal with Graph's object Nodes
 */
import curry from 'lodash/curry';
import flow from 'lodash/flow';
import indexOf from 'lodash/indexOf';
import isString from 'lodash/isString';
import upperFirst from 'lodash/upperFirst';

import throwInDev from './throwInDev';
import { NodeRecord } from '../constants/flowdesigner.model';
import { isPositionElseThrow } from './position';
import { isSizeElseThrow } from './size';
import { Data } from './data';

const positionSelector = ['graphicalAttributes', 'position'];
const sizeSelector = ['graphicalAttributes', 'nodeSize'];
const componentTypeSelector = ['graphicalAttributes', 'nodeType'];

const FORBIDEN_GRAPHICAL_ATTRIBUTES = ['position', 'nodeSize', 'nodeType'];

/**
 * Test if the first parameter is a NodeRecord instance
 * @param {NodeRecord} node
 * @param {bool} doThrow - throw if not a node
 * @returns {bool}
 * @throws
 */
function isNode(node) {
	if (node && node instanceof NodeRecord) {
		return true;
	}
	return false;
}

/**
 * Test if the first parameter is a NodeRecord, throw if not
 * @param {*} node
 * @returns {bool}
 * @throws
 */
export function isNodeElseThrow(node) {
	const test = isNode(node);
	if (!test) {
		throwInDev(
			`Should be a NodeRecord was given ${node &&
				node.toString()}, you should use Node module functions to create and transform Nodes`,
		);
	}
	return test;
}

/**
 * @param {NodeRecord} node
 * @returns {string}
 */
function getId(node) {
	if (isNodeElseThrow(node)) {
		return node.get('id');
	}
	return null;
}

/**
 * @param {string} id
 * @param {NodeRecord}
 * @returns {NodeRecord}
 */
const setId = curry((id, node) => {
	if (isString(id) && isNodeElseThrow(node)) {
		return node.set('id', id);
	}
	throwInDev(`nodeId should be a string was given ${id && id.toString()}}`);
	return node;
});

/**
 * @param {NodeRecord} node
 * @returns {PositionRecord}
 */
function getPosition(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(positionSelector);
	}
	return null;
}

/**
 * @param {PositionRecord} position
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setPosition = curry((position, node) => {
	if (isPositionElseThrow(position) && isNodeElseThrow(node)) {
		return node.setIn(positionSelector, position);
	}
	return node;
});

/**
 * @param {NodeRecord} node
 * @returns {PositionRecord}
 */
function getSize(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(sizeSelector);
	}
	return null;
}

/**
 * @param {SizeRecord} size
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setSize = curry((size, node) => {
	if (isSizeElseThrow(size) && isNodeElseThrow(node)) {
		return node.setIn(sizeSelector, size);
	}
	return node;
});

function getComponentType(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(componentTypeSelector);
	}
	return null;
}

const setComponentType = curry((nodeType, node) => {
	if (isString(nodeType) && isNodeElseThrow(node)) {
		return node.setIn(componentTypeSelector, nodeType);
	}
	throwInDev(`nodeType should be a string was given ${nodeType && nodeType.toString()}`);
	return node;
});

/**
 * @param {String} key
 * @param {any} value
 * @param {nodeRecord} node
 * @returns {nodeRecord}
 */
const setData = curry((key, value, node) => {
	if (isNodeElseThrow(node)) {
		return node.set('data', Data.set(key, value, node.get('data')));
	}
	return node;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
const getData = curry((key, node) => {
	if (isNodeElseThrow(node)) {
		return Data.get(key, node.get('data'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
const hasData = curry((key, node) => {
	if (isNodeElseThrow(node)) {
		return Data.has(key, node.get('data'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const deleteData = curry((key, node) => {
	if (isNodeElseThrow(node)) {
		return node.set('data', Data.delete(key, node.get('data')));
	}
	return node;
});

/**
 * given a key check if that key is white listed
 * @param {String} key
 * @returns {Bool}
 */
function isWhiteListAttribute(key) {
	if (indexOf(FORBIDEN_GRAPHICAL_ATTRIBUTES, key)) {
		return true;
	}
	throwInDev(
		`${key} is a protected value of the Node, please use get${upperFirst(key)} set${upperFirst(
			key,
		)} from this module to make change on those values`,
	);
	return false;
}

/**
 * @param {String} key
 * @param {any} value
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setGraphicalAttribute = curry((key, value, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return node.set(
			'graphicalAttributes',
			Data.set(key, value, node.get('graphicalAttributes')),
		);
	}
	return node;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
const getGraphicalAttribute = curry((key, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return Data.get(key, node.get('graphicalAttributes'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
const hasGraphicalAttribute = curry((key, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return Data.has(key, node.get('graphicalAttributes'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const deleteGraphicalAttribute = curry((key, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return node.set('graphicalAttributes', Data.delete(key, node.get('graphicalAttributes')));
	}
	return node;
});

/**
 * Create a new Node
 * @param {String} id
 * @param {PositionRecord} position
 * @param {SizeRecord} size
 * @param {String} componentType
 * @returns {NodeRecord}
 */
const create = curry((id, position, size, componentType) =>
	flow([setId(id), setPosition(position), setSize(size), setComponentType(componentType)])(
		new NodeRecord(),
	),
);

export const Node = {
	create,
	isNode,
	getId,
	getPosition,
	setPosition,
	getSize,
	setSize,
	getComponentType,
	setComponentType,
	setData,
	getData,
	hasData,
	deleteData,
	setGraphicalAttribute,
	getGraphicalAttribute,
	hasGraphicalAttribute,
	deleteGraphicalAttribute,
};
