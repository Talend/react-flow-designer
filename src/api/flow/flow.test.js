import Immutable from 'immutable';

import * as Node from '../node/node';
import * as Position from '../position/position';
import * as Size from '../size/size';

import * as Flow from './flow';

describe('flow', () => {
	describe('node', () => {
		describe('addNode', () => {
			it('whatever', () => {
				const node = Node.create(
					'nodeId',
					Position.create(100, 150),
					Size.create(70, 120),
					'nodeType',
				);

				expect(Flow.addNode(new Immutable.Map(), node)).toMatchSnapshot();
			});
		});

		describe('updateNode', () => {
			it('whatever', () => {
				const node = Node.create(
					'nodeId',
					Position.create(100, 150),
					Size.create(70, 120),
					'nodeType',
				);
				const updatedNode = Node.create(
					'newNodeId',
					Position.create(100, 150),
					Size.create(70, 120),
					'nodeType',
				);
				const initState = Flow.addNode(new Immutable.Map(), node);

				expect(Flow.updateNode(initState, 'nodeId', updatedNode)).toMatchSnapshot();
			});
		});

		describe('deleteNode', () => {
			it('whatever', () => {
				const node = Node.create(
					'nodeId',
					Position.create(100, 150),
					Size.create(70, 120),
					'nodeType',
				);

				const initState = Flow.addNode(new Immutable.Map(), node);

				expect(Flow.deleteNode(initState, 'nodeId')).toMatchSnapshot();
			});
		});
	});
});
