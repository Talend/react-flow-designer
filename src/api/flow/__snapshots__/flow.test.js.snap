exports[`flow node addNode whatever 1`] = `
Object {
  "childrens": Object {
    "nodeId": Object {},
  },
  "in": Object {
    "nodeId": Object {},
  },
  "nodes": Object {
    "nodeId": Object {
      "data": Object {
        "datasetInfo": Object {},
        "description": "",
        "label": "",
        "properties": Object {},
      },
      "getNodeType": [Function],
      "getPosition": [Function],
      "getSize": [Function],
      "graphicalAttributes": Object {
        "description": "",
        "label": "",
        "nodeSize": Object {
          "height": 120,
          "width": 70,
        },
        "nodeType": "nodeType",
        "position": Object {
          "x": 100,
          "y": 150,
        },
        "properties": Object {},
      },
      "id": "nodeId",
      "type": undefined,
    },
  },
  "out": Object {
    "nodeId": Object {},
  },
  "parents": Object {
    "nodeId": Object {},
  },
}
`;

exports[`flow node deleteNode whatever 1`] = `
Object {
  "childrens": Object {},
  "in": Object {},
  "nodes": Object {},
  "out": Object {},
  "parents": Object {},
}
`;

exports[`flow node updateNode whatever 1`] = `
Object {
  "childrens": Object {
    "newNodeId": Object {},
  },
  "in": Object {
    "newNodeId": Object {},
  },
  "nodes": Object {
    "newNodeId": Object {
      "data": Object {
        "datasetInfo": Object {},
        "description": "",
        "label": "",
        "properties": Object {},
      },
      "getNodeType": [Function],
      "getPosition": [Function],
      "getSize": [Function],
      "graphicalAttributes": Object {
        "description": "",
        "label": "",
        "nodeSize": Object {
          "height": 120,
          "width": 70,
        },
        "nodeType": "nodeType",
        "position": Object {
          "x": 100,
          "y": 150,
        },
        "properties": Object {},
      },
      "id": "newNodeId",
      "type": undefined,
    },
  },
  "out": Object {
    "newNodeId": Object {},
  },
  "parents": Object {
    "newNodeId": Object {},
  },
}
`;
